draft-ietf-tls-tls13.md

```diff
@@ -1585,28 +1582,25 @@ exist). As with ServerHello, a
 HelloRetryRequest MUST NOT contain any extensions that were not first
 offered by the client in its ClientHello.
 
-Upon receipt of a HelloRetryRequest, the client MUST first verify that
-the selected_group field corresponds to a group which was provided
-in the "supported_groups" extension in the original ClientHello.  It
-MUST then verify that the selected_group field does not correspond
-to a group which was provided in the "key_share" extension in the
-original ClientHello. If either of these checks fails, then the client
-MUST abort the handshake with a fatal "illegal_parameter"
-alert. Clients SHOULD also abort with "unexpected_message" in response
-to any second HelloRetryRequest which was sent in the same connection
-(i.e., where the ClientHello was itself in response to a
-HelloRetryRequest).
-
-Otherwise, the client MUST send a ClientHello with an updated KeyShare
-extension to the server. The client MUST append a new KeyShareEntry
-for the group indicated in the selected_group field to the groups
-in its original KeyShare.
-
-Upon re-sending the ClientHello and receiving the server's
-ServerHello/KeyShare, the client MUST verify that the selected
-NamedGroup matches that supplied in the HelloRetryRequest and MUST
-abort the connection with a fatal "illegal_parameter" alert if it does
-not.
+Upon receipt of a HelloRetryRequest, the client MUST verify that the extensions
+block is not empty and otherwise abort the handshake with a fatal
+"decode_error" alert.  Clients SHOULD also abort with "unexpected_message" in
+response to any second HelloRetryRequest which was sent in the same connection
+(i.e., where the ClientHello was itself in response to a HelloRetryRequest).
+
+Otherwise, the client MUST process all extensions in the HelloRetryRequest and
+send a second updated ClientHello. The HelloRetryRequest extensions defined in
+this specification are:
+
+- cookie [{{cookie}}]
+
+- key_share [{{key-share}}]
+
+Note that HelloRetryRequest extensions are defined such that the original
+ClientHello may be computed from the new one, given minimal state about which
+HelloRetryRequest extensions were sent. For example, the key_share extension
+causes the new KeyShareEntry to be appended to the client_shares field, rather
+than replacing it.
```

Comment:
I think that it would help to point out here that extensions in HelloRetryRequest can be sent by the server even if the client did not include them in their ClientHello.  Or maybe we just need a special rule for `cookie`.